### PR TITLE
Apply logical type preparation when checking for best union type

### PR DESCRIPTION
**Pull request by ksunden**
Previously validate would pass (and it does the prepare step) but then the datum was still pre preparation when it tried to do anything further. 
As far as I can tell this only really matters for record types, but made more general in case there was an edge case I was missing. 

The provided test shows this for an ndarray record type which would fail because set(ndarray) does not work [well, it actually does for 1D arrays, but not 2+ dimensional arrays... even in the 1d case it was doing something funky that technically worked, but only on accident]

### CODE REVIEW

**Comment by scottbelden** [comment]
This is an interesting use case of a `logicalType`. None of the logical types currently defined in the Avro specification use `record` as the base type. Most use a primitive type, but `duration` uses `fixed` which is also a named type. The specification states that `A logical type is an Avro primitive or complex type with extra attributes to represent a derived type.` so I don't think there is any reason to say that a record can't be a logical type.

I think something like this makes sense, but I'm going to think about it a little and get back to you.

File: fastavro/_write.pyx [hunk 1]
```diff
@@ -295,6 +295,13 @@ cdef write_union(bytearray fo, datum, schema, dict named_schemas, fname):
                     continue
 
             if _validate(datum, candidate, named_schemas, raise_errors=False):
+                if isinstance(candidate, dict):
```
Comment (scottbelden) [inline fastavro/_write.pyx:298]: I saw the part about the check being done here for the general case, but I think we can safely keep this logic in the specific record case. Therefore, could you move this down under the `if record_type == "record":` block? We can also get rid of the `if isinstance(candidate, dict):` check at that point.

File: fastavro/_write_py.py [hunk 2]
```diff
@@ -183,6 +183,13 @@ def write_union(encoder, datum, schema, named_schemas, fname):
                     continue
 
             if _validate(datum, candidate, named_schemas, raise_errors=False):
+                if isinstance(candidate, dict):
```
Comment (scottbelden) [inline fastavro/_write_py.py:186]: Same comment as for the other `write` file.

File: tests/test_logical_types.py [hunk 3]
```diff
@@ -474,3 +475,80 @@ def test_local_timestamp_micros():
...
+
+
+class Interface:
+    def __init__(self, array_interface):
+        array_interface["shape"] = tuple(array_interface["shape"])
+        self.__array_interface__ = array_interface
+
+
+def read_ndarray(data, writer_schema, reader_schema):
+    import numpy as np
```
Comment (scottbelden) [inline tests/test_logical_types.py:487]: This can be removed since the import is at the top of the file.

**Review by scottbelden — CHANGES_REQUESTED** [review]
I think this makes sense. Just a few small comments and then it should be good to go. Thanks!

File: tests/test_logical_types.py (hunk 3 above)
Comment (ksunden) [reply tests/test_logical_types.py]: ah, this was copy pasted from some code that I wanted to register the logical type conversion, but not import numpy until it actually needed it

**Review by scottbelden — APPROVED** [review]
